From f8d7aecf9d878ab4f72dbca6b17914df9c167b29 Mon Sep 17 00:00:00 2001
From: Yimin Peng <quic_yiminp@quicinc.com>
Date: Tue, 8 Aug 2023 16:59:15 +0800
Subject: [PATCH 2/3] kiumd: per-process pagetable for kgsl

Enabling TTBR1 for global pagetable and setting TTBRO for
per process page table. Allocate a pagetable for a process
and return the pagetable physical address and asid for GPU
to program the ttbr0 to switch pagetables.

Change-Id: I8fd9b0a518c2bce9c5773785d012bc37caaa9de4
Signed-off-by: Aravind Vijayakumar <quic_aprasann@quicinc.com>
Signed-off-by: Yimin Peng <quic_yiminp@quicinc.com>
---
 drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c | 21 ++++++++++++---------
 include/linux/adreno-smmu-priv.h           |  2 ++
 include/linux/qcom_scm.h                   |  1 +
 3 files changed, 15 insertions(+), 9 deletions(-)

diff --git a/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c b/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c
index 501bb3f13039..d93d18543f59 100644
--- a/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c
+++ b/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c
@@ -427,7 +427,7 @@ static const struct io_pgtable_cfg *qcom_adreno_smmu_get_ttbr1_cfg(
  * are active
  */
 
-static int qcom_adreno_smmu_set_ttbr0_cfg(const void *cookie,
+int qcom_adreno_smmu_set_ttbr0_cfg(const void *cookie,
 		const struct io_pgtable_cfg *pgtbl_cfg)
 {
 	struct arm_smmu_domain *smmu_domain = (void *)cookie;
@@ -467,6 +467,7 @@ static int qcom_adreno_smmu_set_ttbr0_cfg(const void *cookie,
 
 	return 0;
 }
+EXPORT_SYMBOL(qcom_adreno_smmu_set_ttbr0_cfg);
 
 static int qcom_adreno_smmu_alloc_context_bank(struct arm_smmu_domain *smmu_domain,
 					       struct arm_smmu_device *smmu,
@@ -529,11 +530,12 @@ static int qcom_adreno_smmu_init_context(struct arm_smmu_domain *smmu_domain,
 	 */
 
 	priv = dev_get_drvdata(dev);
-	priv->cookie = smmu_domain;
-	priv->get_ttbr1_cfg = qcom_adreno_smmu_get_ttbr1_cfg;
-	priv->set_ttbr0_cfg = qcom_adreno_smmu_set_ttbr0_cfg;
-	priv->get_fault_info = qcom_adreno_smmu_get_fault_info;
-	priv->pgtbl_info = *input_info;
+	if(priv != NULL) {
+		priv->cookie = smmu_domain;
+		priv->get_ttbr1_cfg = qcom_adreno_smmu_get_ttbr1_cfg;
+		priv->set_ttbr0_cfg = qcom_adreno_smmu_set_ttbr0_cfg;
+		priv->get_fault_info = qcom_adreno_smmu_get_fault_info;
+		priv->pgtbl_info = *input_info;
 
 	/*
 	 * These functions are only compatible with the data structures used by the
@@ -543,9 +545,10 @@ static int qcom_adreno_smmu_init_context(struct arm_smmu_domain *smmu_domain,
 	 * Providing these functions as part of the GPU interface also makes little sense
 	 * as context banks are set to stall by default anyway.
 	 */
-	if (of_match_node(qcom_smmu_impl_of_match, np)) {
-		priv->set_stall = qcom_adreno_smmu_set_stall;
-		priv->resume_translation = qcom_adreno_smmu_resume_translation;
+		if (of_match_node(qcom_smmu_impl_of_match, np)) {
+			priv->set_stall = qcom_adreno_smmu_set_stall;
+			priv->resume_translation = qcom_adreno_smmu_resume_translation;
+		}
 	}
 
 	return 0;
diff --git a/include/linux/adreno-smmu-priv.h b/include/linux/adreno-smmu-priv.h
index 376e65b7fb3d..076d403c35f1 100644
--- a/include/linux/adreno-smmu-priv.h
+++ b/include/linux/adreno-smmu-priv.h
@@ -10,6 +10,8 @@
 #include <linux/io-pgtable.h>
 #include <linux/qcom-io-pgtable.h>
 
+int qcom_adreno_smmu_set_ttbr0_cfg(const void *cookie, const struct io_pgtable_cfg *pgtbl_cfg);
+
 /**
  * struct adreno_smmu_fault_info - container for key fault information
  *
diff --git a/include/linux/qcom_scm.h b/include/linux/qcom_scm.h
index c580ca808e88..1ff7c78250db 100644
--- a/include/linux/qcom_scm.h
+++ b/include/linux/qcom_scm.h
@@ -235,6 +235,7 @@ extern int qcom_scm_ocmem_lock(enum qcom_scm_ocmem_client id, u32 offset,
 			       u32 size, u32 mode);
 extern int qcom_scm_ocmem_unlock(enum qcom_scm_ocmem_client id, u32 offset,
 				 u32 size);
+extern int qcom_scm_kgsl_set_smmu_aperture(unsigned int num_context_bank);
 
 extern int qcom_scm_config_set_ice_key(uint32_t index, phys_addr_t paddr,
 				       size_t size, uint32_t cipher,
-- 
2.17.1

